gatsby/content/blog/2021/12/2021-12-10-sydent-typing-2.mdx

```diff
@@ -0,0 +1,387 @@
+```
+
+I could more correctly annotate the `YieldType` as `Deferred[Union[_RoutingResult, IResponse]]` so that the `SendType` was `Union[_RoutingResult, IResponse]`. But this would mean we end up having some kind of type check at each yield point: a `cast`, or a `type: ignore`, or a runtime `isinstance` check. It didn't feel like it was worth the boilerplate, especially since I could narrow e.g. `res: Any` to `res: IResponse` with minimal effort.
+
+This problem doesn't arise with using the more common `async/await` syntax:
+
+```python
+async def foo() -> int:
+    return 1
+
+async def bar() -> None:
+    x = await foo()
+    reveal_type(foo())
+    reveal_type(x)
+```
+
+```
+$ mypy example.py
+example.py:6: note: Revealed type is "typing.Coroutine[Any, Any, builtins.int]"
+example.py:7: note: Revealed type is "builtins.int*"
+```
+
+All in all, the handling of `inleineCallbacks` is a situation specific to working with (older?) twisted code. It's still nice to understand what's going on behind the scenes though!
+
+
+
+#### `zope.interface.Interface`
+
+Twisted makes use of `zope`'s `Interface` to define a number of abstract interface classes. Speaking personally, I've not seen it used outside twisted, and I think that means it's not supported by much of the typechecking tooling. For example, I've definitely seen PyCharm struggle to realise that it's okay to pass a `Response` to a function which expects an `IResponse`! Here's a more complicated example where `PyCharm` isn't happy with me widening the type `LoggingHostnameEndpoint` to `IStreamClientEndpoint`, even though [the latter implements the former](https://github.com/matrix-org/sydent/blob/92ff7a878a25696365b10cc49e32f5cba32c5960/sydent/http/matrixfederationagent.py#L379-L380).
+
+![Screenshot from pycharm showing a false positive warning](https://i.imgur.com/27rqjHm.png)
+
+
+Mypy out of the box doesn't play well with a zope `Interface` (nor does any other typechecker I tried). Fortunately, the excellent [`mypy-zope`](https://github.com/Shoobx/mypy-zope) plugin helps here: it teaches mypy that any class like `Response` which `@implements(IResponse)`  can be passed in place of an `IResponse`.
+
+#### You're at the mercy of your dependencies
+
+Hmm. That's being a little melodramatic.
+
+Improving coverage boils down to giving the typechecker more information about your program. The more information it has, the more it can check—and the more errors it can spot. (Hopefully this doesn't make typing come across like a pyramid scheme.) If your dependencies aren't typed, mypy can't validate you're correctly providing inputs and correctly consuming outputs. You might have a bigger impact on overall typing coverage by annotating a dependency (directly or via stubs). I have a hunch that bugs are more likely in code that uses an external dependency: we're much more familiar with the details of our own source code compare to that of a third party we trust.
```

Comment:
```suggestion
Improving coverage boils down to giving the typechecker more information about your program. The more information it has, the more it can check—and the more errors it can spot. (Hopefully this doesn't make typing come across like a pyramid scheme.) If your dependencies aren't typed, mypy can't validate you're correctly providing inputs and correctly consuming outputs. You might have a bigger impact on overall typing coverage by annotating a dependency (directly or via stubs). I have a hunch that bugs are more likely in code that uses an external dependency: we're much more familiar with the details of our own source code compared to that of a third party we trust.
```